### ISSUE TITLE
Enable a mechanism to capture coordinate reference system metadata

### ISSUE THREAD

**Issue by mwtoews** [MODFLOW-ORG/modflow6#1323]
This is a feature request to enable a MF6 model to capture information for a coordinate references system (CRS; also called spatial reference system or SRS), so this information can be used by GIS software and tools. This information would not be used directly by MODFLOW 6, but could potentially be passed to the binary grid file (.grb).

### Background
There are several mechanisms that software can capture CRS info. The simplest method is to use look-up authority names and codes. The most commonly used authority is [EPSG](https://epsg.org), but several authorities also exist. For example, using pyproj:
```
$ python -c "import pyproj; print(pyproj.database.get_authorities())"
['EPSG', 'ESRI', 'IAU_2015', 'IGNF', 'NKG', 'OGC', 'PROJ']
```
and the second piece is the code, which is always an integer greater than one. For example, here are some projected CRSs on Earth:

- `EPSG:26916` - [NAD83 / UTM zone 16N](https://epsg.org/crs_26916/index.html)
- `ESRI:102007` - [NAD 1983 Albers Hawaii](https://epsg.io/102007)
- `IAU:2015:39916` - [Earth (2015) / Ographic/ Equirectangular, clon = 180](http://www.opengis.net/def/crs/IAU/2015/39916)

Software like PyPROJ can identify these CRS like this:
```python
from pyproj import CRS
crs1 = CRS.from_authority("EPSG", 26916)
crs2 = CRS.from_authority("ESRI", 102007)
crs3 = CRS.from_authority("IAU_2015", 39916)
```
and Esri's proprietary ArcPy has a [SpatialReference](https://pro.arcgis.com/en/pro-app/latest/arcpy/classes/spatialreference.htm) interface:
```python
import arcpy
sr1 = arcpy.SpatialReference(26916)
sr2 = arcpy.SpatialReference(102007)
```
(according to their [docs](https://www.esri.com/arcgis-blog/products/arcgis-pro/mapping/coordinate-systems-difference/), the WKID code will either refer to EPSG or ESRI, but there is no need to specify the authority since the codes don't overlap.)

New authority names and codes are added/updated every year, so capturing this data is only as good as the client software is able to look up the information (e.g. older software won't understand future codes).

An alternative mechanism to capture CRS is to explicitly capture the coordinate reference types and parameters, which is complicated (e.g. WKT1, [WKT2:2019](https://docs.ogc.org/is/18-010r7/18-010r7.html), [PROJJSON](https://proj.org/specifications/projjson.html), [CF Metadata Conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#coordinate-system), etc.). This complex approach should be avoided due to burden of information that would need to be handled, and the complexity of different markup schemes.

This approach would limit users to only use registered CRS codes. Custom or local CRSs would not be permitted.

### Discussion of idea
A suggestion for discritization input option blocks is to be modified as follows:
```
BEGIN OPTIONS
  [LENGTH_UNITS <length_units>]
  [NOGRB]
  [CRS_DEF <crs_def>]
  [XORIGIN <xorigin>]
  [YORIGIN <yorigin>]
  [ANGROT <ang>]
END OPTIONS
```
This would describe the short CRS definition for XORIGIN/YORIGIN, but have no other relation with LENGTH_UNITS or ANGROT, which are local coordinates used by MODFLOW 6.

The new option CRS_DEF would be a string with some valid entry examples:
```
EPSG:26916
ESRI:102007
IAU:2015:39916
FUTURE:123
```
invalid entries would be missing an authority name (e.g. `26916`), missing ":" separator (e.g. `EPSG_26916`), or have non-integer authority code (e.g. `EPSG:26916A`).

Using some Python pseudocode, this data could be stored as two entities:
```python
crs_def = "IAU:2015:39916"
crs_auth_name, crs_auth_code = crs_def.rsplit(":", 1)
crs_auth_code = int(crs_auth_code)
print(f"{crs_auth_name=}, {crs_auth_code=}")
# crs_auth_name='IAU:2015', crs_auth_code=39916
```
within Fortran, `crs_auth_name` could be a character string with max length 10 and `crs_auth_code` would be an integer. Or `crs_def` could be stored as a single character string with max length 20.

Unsure if/how this would make it into the binary grid specification. The current is "VERSION 1", so if a next version is planned, this information could be stored in the .grb file. I've also noticed that the format specification is missing LENGTH_UNITS too.

**Comment by langevin-usgs** [comment]
@mjreno, drawing your attention to this issue as @mwtoews has thoughts on how to specify the model CRS.  Not sure this is consistent with your NetCDF work.

**Comment by aleaf** [comment]
@mwtoews @christianlangevin @wpbonelli  I second this idea as it would finally make MODFLOW models self-describing in terms of their location in space. To provide a concrete example, currently when models are loaded using Flopy, a modelgrid instance with real-world coordinates is created automatically. But the user still has to assign a CRS manually in a script:
```python
sim = flopy.mf6.MFSimulation.load(sim_name, sim_ws=model_ws)
gwf = sim.get_model(model_name)
gwf.modelgrid.crs = 26915  # the user has to know this and input it correctly
```
Assigning a CRS is important because without it, any GIS files (shapefiles, rasters, geopackages, etc) exported from the model likely won't plot correctly in a GIS environment, unless the environment is set to the correct CRS (that xorigin and yorigin are referenced to). A CRS is also required for any coordinate reprojection, for example mapping data referenced in another coordinate system to the model grid. Coordinate reprojection is a routine modeling task that is best automated; not having the CRS stored with the model is a sticking point.

I would just add two suggestions to what @mwtoews has above:
1) Just use `CRS` instead of `CRS_DEF` as the variable name in the DIS Package file.
2) The input shouldn't have to conform to any particular format. In principal, any input that describes a valid CRS should be allowed. As @mwtoews notes, MODFLOW itself doesn't have to care about this. Flopy could validate CRS input by simply passing it to the [pyproj CRS constructor](https://pyproj4.github.io/pyproj/stable/api/crs/crs.html), and throwing an error or warning if it doesn't result in a valid `CRS` instance.

**Comment by mjreno** [comment]
@mwtoews @aleaf 
Apologies for not following up on this suggestion earlier. I also like the idea and in keeping with the thought that this parameter would be transparent to MODFLOW 6 otherwise I agree that there wouldn't need to be restrictions on the input itself aside from a max length requirement so it could be written to grb file. The input validation would come in post processing environments.

It seems like a CRS header field could be added to the grb file- not sure what sort of backwards compatibility issues might arise by extending it in this way. This information could also be written to MODFLOW 6 model NetCDF output files, perhaps as a global attribute with no other associations with the internal data.

As an aside, MODFLOW 6 does have a NetCDF configuration (UTL-NCF) package, which allows a user to specify a WKT string which is written to the model NetCDF file. In this case grid variables are associated with a mapping variable, the intention being that NetCDF files can then be visualized in GIS using the expected coordinates. 

Adding the input parameter will be straightforward- I'll look into potential ramifications related to extending the grb format. I'll add that here when I have it. Thanks for the suggestion happy to hear more input if you have it.

**Comment by aleaf** [comment]
thanks @mjreno! I'm happy to follow up on the Flopy side too once you get something put together.

**Comment by mwtoews** [comment]
Another "cheaper" idea to implement this, but (by my perspective) less preferred, is to create another [sidecar file](https://en.wikipedia.org/wiki/Sidecar_file)  `model.dis.prj` file containing a WKT of projection information. This is similar to shapefile's `file.prj`, but also similar to `model.dis.grb`. This ideal wouldn't require any modification with MODFLOW's file spec, as it would be an extra file. It would only be a convention that software like FloPy would need to be aware of.

**Comment by mjreno** [comment]
@mwtoews I was thinking something similar- a file analogous to the usgs.model.reference files.  Here's an example:
```
xul 1954815.00          # x-coordinate of the upper left of the model grid
yul 219671.000          # y-coordinate of the upper left of the model grid
rotation  18            # rotation of the model grid in degrees
length_units feet       # length unit of the model grid (feet, meters, etc.)
time_units days         # time unit of the model grid (days, years, etc.)
start_date 01/01/2020   # state date of the model
start_time 00:00:00     # start time of the model
model MODFLOW 6         # MODFLOW model type (MODFLOW-NWT, etc.)
EPSG 4456               # EPSG code
```
This could be modified to suit the purpose, including a more generic "CRS" user input string.  Perhaps the ```prj``` file you are suggesting is more appropriate if only the projection info is needed.

Another possibility is introducing a new grb version (e.g. ```1.1``` or ```2```) that is only associated with the extra CRS parameter.  MODFLOW 6 would continue to produce version 1 grb files when the optional input is not provided.

**Comment by aleaf** [comment]
I would agree that a sidecar file is less preferred. More clutter, more potential for it to get lost. It's also separate from the xorigin and yorigin specification in the DIS file, so there's potential for stale files causing problems. Maybe still worth supporting outside of MODFLOW for locating older models, or when older versions of MODFLOW 6 must be used. 

One problem we've had in the past is multiple (some admittedly half-baked) ways to input CRS-related information, including the above usgs.model.reference in addition to the namefile header comments, DIS package x,y origin in MODFLOW 6 and length units specifier in older versions of MODFLOW, as well as grid spec in the PEST world. Which led to a lot of confusion, especially among modelers that weren't as knowledgeable about GIS. I can't tell you how many models I've come across that have been difficult to locate because of incorrect or conflicting information.

I guess part of the appeal here is that the current information in the MODFLOW 6 DIS package is pretty clean, simple and almost complete. It just needs this CRS piece.
 
I must confess I'm not well informed on the implications of modifying the grb file, so don't have strong opinions there. Is it a big deal to add an optional CRS argument to the DIS package options block?

**Comment by mjreno** [comment]
Getting the new parameter into the package isn't a big deal but updating the binary format will break other applications that directly parse the version 1 format.  A sidecar file sidesteps the issue of updating the version 1 ```grb``` format.  Supporting multiple versions perhaps introduces more clutter and confusion.

One other thought is a NetCDF version of the ```grb``` file.  This wouldn't suffer from the same format restrictions as the binary file but would require some new tooling and dependencies.

**Comment by christianlangevin** [comment]
Just an FYI that it would be possible to add CRS information to the binary grid file without breaking backward compatibility.  The GRB specification is somewhat flexible, as shown below.  There can be any number of "Definition 0" comment entries as long as they start with the # comment character.  The lines must not be greater than LENTEXT (100 at the moment), but it seems like we could put that coordinate reference information into the GRB file.

![Image](https://github.com/user-attachments/assets/8681d019-390a-48c5-9b1d-2f4fc23638ad)

**Comment by mjreno** [comment]
I think that would limit the types of user inputs that would be supported but is probably a good compromise for a first step.  I'd be interested to hear from @aleaf if the 100 character limitation would be sufficient to cover most practical modeling scenarios.

**Comment by mjreno** [comment]
If it sounds reasonable then I'm thinking to add a ```DIS CRS``` parameter and a new ```FILEOUT``` option for the projection file.  If ```CRS``` is provided with the ```prj FILEOUT``` option the provided ```CRS``` info will be written there.  If ```CRS``` is provided without the ```prj FILEOUT``` option it will be written to the ```grb``` file comment field if it fits the length restriction, and if not this will generate an error.

**Comment by mwtoews** [comment]
Based on the discussion, here are my takes:

1. Add a new `CRS <crs_user_input>` option for each DIS package for user input of 100 chars. This should be treated as case-sensitive. PROJ and related (pyproj) can instantiate a CRS from a few [user inputs](https://pyproj4.github.io/pyproj/stable/api/crs/crs.html#pyproj.crs.CRS.from_user_input), but I suggest promoting a few "common" ones like
   1. Authority string of the format `AUTHORITY:CODE`, e.g. [`EPSG:25832`](https://spatialreference.org/ref/epsg/25832/), [`ESRI:102007`](https://spatialreference.org/ref/esri/102007/)
   2. An OGC URN (Universal Resource Name), e.g. `urn:ogc:def:crs:EPSG::25832`
   3. A PROJ.4 string, which is least preferred because it is lossy/deprecated. These are _usually_ less than 100 chars, e.g. `+proj=utm +zone=32 +ellps=GRS80 +units=m +no_defs +type=crs`
2. If a CRS option is defined, write that to the GRB file in the comment: `# CRS: <crs_user_input>`

I don't think mf6 needs to otherwise be concerned with reading/writing `.prj` files, as they are not relevant to groundwater simulation. The other reason why I'm less enthusiastic about adding a `.prj` sidecar file is that it may contain one of several WKT formats (WKT2:2015, WKT2:2019, WKT1:GDAL, WKT1:ESRI).

**Comment by mjreno** [comment]
I can see it both ways. I agree that these are import considerations but I'm not sure that MODFLOW 6 is the right context to promote or regulate best practices related to the CRS specification.  I see that more the responsibility of the pre- and post-processing environments, e.g. FloPy.  From that perspective I'm more concerned with supporting the widest range of user inputs as is reasonable, even if they don't adhere to current best practice.

If there is agreement to restrict ourselves to the binary grid file we can promote @mwtoews list as part of the description in mf6io but I don't think we impose any real restrictions other than length.

**Comment by aleaf** [comment]
Agreed that we don't need to promote any particular formats. But I think it would be helpful in the mf6 IO instructions to give a few common examples so that users know how to get this to work. And I think those examples should be best practices that would be easily parsed by Flopy or other post-processors. Basically it has to be a string that Flopy or another tool can pass as-is to pyproj/PROJ, without any manipulations.

re: using a sidecar file for WKT- I was thinking that the sidecar file would be provided by the user, and just referenced in the DIS package file, so that Flopy, for example would have an easy way to identify the sidecar file and parse it automatically on load.

Any example of what the IO documentation could look like:
```
BEGIN OPTIONS
   [CRS FILEIN <prj_filename>]
END OPTIONS
```
CRS—Real-world coordinate reference system for the model, for example, an EPSG integer code (e.g. 26915) or authority string (i.e. epsg:26915). Limited to 100 characters. Longer CRS representations can be supplied in an external projection file with the optional FILEIN keyword. The entry for CRS does not affect the model simulation, but it is written to the binary grid file so that postprocessors can locate the grid in space.
FILEIN—optional keyword to specify that an input filename is expected next.
prj_filename—defines an external projection file with coordinate reference information; for example, a *.prj file containing Well-known text like those used for shapefiles.

or if it's cleaner to have a separate variable for the external projection file:

```
BEGIN OPTIONS
   CRS
   [PRJ FILEIN <prj_filename>]
END OPTIONS
```

CRS—Real-world coordinate reference system for the model, for example, an EPSG integer code (e.g. 26915) or authority string (i.e. epsg:26915). Limited to 100 characters. Longer CRS representations can be supplied in an external projection file with the optional PRJ FILEIN keywords. The entry for CRS does not affect the model simulation, but it is written to the binary grid file so that postprocessors can locate the grid in space.
PRJ—keyword to specify that the record corresponds to an external projection file with coordinate reference system information.
FILEIN—optional keyword to specify that an input filename is expected next.
prj_filename—defines an external projection file with a coordinate reference information, for example, a *.prj file containing Well-known text like those used for shapefiles.

**Comment by mjreno** [comment]
That's an interesting thought @aleaf. Would the file always contain a single line of text with a prescribed maximum length, i.e. 100? Do you see distinct workflow use cases for the CRS string in the projection file and binary grid file? If that is not currently clear it would be a trivial first step to support the user provided projection file and then later to write that string to the binary grid file when it is clearly relevant.

This also would give us some time to consider the best way to get this information into the grid file.  At this point it seems clear that even adding comments will require some adjustments in certain consumers of the grid file.  If we need to do these updates anyways it might make more sense to support a new version of the file (along with the current one) that more fully supports configurable string fields.  In this case we wouldn't necessarily be limited to 100 characters but we would still need to define a character limit (or set of limits) for these fields in the file.  So it would be good to think about what limit we would place on the CRS input string if we could define it.

**Comment by mwtoews** [comment]
I still don't see why mf6 would need to handle any external projection file. These needs/cases should be handled separately by client applications, i.e. if a user needs to have a model.dis.prj file for their uses. (There are many examples of these, like gridspec files for pest utilities, etc.)

But to the main point of embedding a CRS user input to both the DIS options block and GRB output file, I have a few related thoughts. First on the practical character limit. The GRB file there will need to be at least a 5 character prefix `#CRS:` thus `<crs_user_input>` is limited to a 95 character limit.

If the CRS user input is an externally registered projection, e.g. by EPSG or ESRI authorities, these will easily fit in 95 chars (even with an OGC URN-style definition).

To handle cases where a bespoke CRS is needed, it would probably need to be a one-line PROJ.4 string, which can sometimes be longer than 95 characters. For example, if [EPSG:8316](https://spatialreference.org/ref/epsg/8316) were not previously registered, then a users' DIS file might contain:
```
BEGIN OPTIONS
  CRS +proj=lcc +lat_1=45.5833333333333 +lat_0=45.5833333333333 +lon_0=-123.416666666667 +k_0=1.000045 +x_0=30000.00001512 +y_0=19999.99999992 +ellps=GRS80 +units=ft +no_defs +type=crs
  LENGTH_UNITS feet
  ...
END OPTIONS
```
For this instance, the crs_user_input is 178 chars. I would expect the GRB to repeat the `#CRS:` comment as many times as necessary:
```
#CRS:+proj=lcc +lat_1=45.5833333333333 +lat_0=45.5833333333333 +lon_0=-123.416666666667 +k_0=1.00004
#CRS:5 +x_0=30000.00001512 +y_0=19999.99999992 +ellps=GRS80 +units=ft +no_defs +type=crs
```
Then the client app would combine these back into a 178 char crs_user_input.

**Comment by mjreno** [comment]
I had a similar thought @mwtoews - splitting up long strings into either the comment or dedicated text fields.  @aleaf is this a general enough solution if one way or the other any CRS strings can be represented in or reconstructed from the binary grid file?

**Comment by mjreno** [comment]
@mwtoews @aleaf feel free to review or comment on the current solution in the PR's above.  I'm thinking to merge this soon and then the basic capability would be available in the nightly build.  Flopy has basic support (the CRS string would be read into the internal ```binarygrid_util``` dict) but will need additional work.  @aleaf are you interested in looking at that?

**Comment by aleaf** [comment]
cool, thanks @mjreno! I'll take a look at these.

**Comment by aleaf** [comment]
@mjreno this looks good. I can start working on the additional Flopy support.
